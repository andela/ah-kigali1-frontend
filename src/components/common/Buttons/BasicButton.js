import React from "react";
import PropTypes from "prop-types";
const BasicButton = props => {
  return (
    <button
      type="button"
      className={`btn ${props.className}`}
      {...props}
      onClick={props.onClick}
    >
      {props.title}
    </button>
  );
};

BasicButton.prototype = {
  title: PropTypes.string,
  onClick: PropTypes.func,
  className: PropTypes.string
};
BasicButton.defaultProps = {
  title: ""
};
export default BasicButton;
