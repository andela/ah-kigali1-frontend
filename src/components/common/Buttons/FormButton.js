import React from "react";
import PropTypes from "prop-types";
const FormButton = props => {
  return (
    <input
      {...props}
      type="button"
      value={props.value}
      onClick={props.onClick}
      disabled={props.disabled}
    />
  );
};

FormButton.prototype = {
  value: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool
};
FormButton.defaultProps = {
  value: "Submit",
  disabled: false
};
export default FormButton;
