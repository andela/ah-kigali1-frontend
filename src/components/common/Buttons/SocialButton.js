import React from "react";
import PropTypes from "prop-types";

const SocialButton = props => {
  const { className, iconName, alt, onClick } = props;
  return (
    <div className={`icon ${className}`} onClick={onClick}>
      <img
        src={require(`../../../assets/icons/${iconName}-icon.svg`)}
        alt={alt}
      />
    </div>
  );
};

SocialButton.propTypes = {
  className: PropTypes.string,
  iconName: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired
};
SocialButton.defaultProps = {
  className: ""
};
export default SocialButton;
