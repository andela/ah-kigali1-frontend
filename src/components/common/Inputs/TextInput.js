import React from "react";
import PropTypes from "prop-types";

const TextInput = ({
  name,
  type,
  id,
  placeholder,
  value,
  onChange,
  ...props
}) => (
  <input
    {...props}
    name={name}
    type={type}
    id={id}
    placeholder={placeholder}
    value={value}
    onChange={onChange}
  />
);

TextInput.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  id: PropTypes.string,
  placeholder: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: PropTypes.string,
  onKeyDown: PropTypes.func
};

TextInput.defaultProps = {
  id: "",
  value: "",
  onKeyDown: () => ""
};

export default TextInput;
