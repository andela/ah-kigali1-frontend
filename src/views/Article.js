import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import addArticle from "../redux/actions/index";
// import "../index.css";

const mapDispatchToProps = dispatch => ({
  addArticle: () => dispatch(addArticle())
});

const FetchedArticles = props => (
  <div>
    <h1>Fetch some articles from dummy app</h1>

    <button
      // eslint-disable-next-line react/destructuring-assignment
      onClick={() => props.addArticle()}
      className="btn"
      type="submit"
    >
      Fetch more articles
    </button>
  </div>
);

FetchedArticles.propTypes = {
  // eslint-disable-next-line react/require-default-props
  addArticle: PropTypes.func
};

const Articles = connect(
  null,
  mapDispatchToProps
)(FetchedArticles);

export default Articles;
