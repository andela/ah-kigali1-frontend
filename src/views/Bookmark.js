import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import BookmarkComponent from "../components/common/Bookmarks";
import bookmarkIcon from "../assets/img/bookmark-icons.svg";
import bookmarkedIcon from "../assets/img/bookmarked.svg";
import { handleBookmark } from "../redux/actions/bookmarkActions";

export class Bookmark extends Component {
  static chooseBookmarkIcon(condition) {
    return condition ? bookmarkedIcon : bookmarkIcon;
  }

  render() {
    const {
      bookmark,
      isBookmarked,
      slug,
      isSubmitting,
      bookmarkedArticles,
      location,
      history
    } = this.props;
    if (!Object.prototype.hasOwnProperty.call(bookmarkedArticles, slug)) {
      Object.assign(bookmarkedArticles, { [slug]: isBookmarked });
    }
    const theBookmarkIcon = Bookmark.chooseBookmarkIcon(
      bookmarkedArticles[slug]
    );
    return (
      <BookmarkComponent
        onClick={() =>
          bookmark(slug, bookmarkedArticles[slug], { location, history })
        }
        icon={theBookmarkIcon}
        disabled={isSubmitting}
      />
    );
  }
}

export const mapStateToProps = state => ({
  ...state.bookmark
});

export const mapDispatchToProps = dispatch => ({
  bookmark: (slug, isBookmarked, { history, location }) =>
    dispatch(handleBookmark(slug, isBookmarked, { history, location }))
});

Bookmark.propTypes = {
  bookmark: PropTypes.func.isRequired,
  isBookmarked: PropTypes.bool.isRequired,
  bookmarkedArticles: PropTypes.shape({}).isRequired,
  slug: PropTypes.string.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  location: PropTypes.shape([]).isRequired,
  history: PropTypes.shape({}).isRequired
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Bookmark)
);
