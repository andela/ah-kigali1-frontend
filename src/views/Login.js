import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Link, Redirect } from "react-router-dom";
import {
  handleTextInput,
  handleSignIn,
  socialAuth
} from "../redux/actions/loginActions";
import { isEmpty } from "../utils/helperFunctions";
import Validator from "../utils/validator";
import TextInput from "../components/common/Inputs/TextInput";
import FormButton from "../components/common/Buttons/FormButton";
import BasicButton from "../components/common/Buttons/BasicButton";
import SocialButton from "../components/common/Buttons/SocialButton";
import Logo from "../components/common/Logo";

export class Login extends Component {
  state = {
    errors: {}
  };

  handleOnChange = e => {
    this.setState({
      errors: {}
    });
    const { name, value } = e.target;
    const { handleTextInput: textInput } = this.props;
    textInput(name, value);
  };

  handleSubmit = () => {
    const { email, password, handleSignIn: signIn } = this.props;
    const errors = Validator.formData({ email, password });
    if (!isEmpty(errors)) {
      this.setState({
        errors: {
          ...errors
        }
      });
      return;
    }
    signIn({ email, password });
  };

  handleNavigation = path => <Redirect to={`/${path}`} />;

  handleSocialAuth(provider) {
    const { isSubmitting, socialAuth: socialLogin } = this.props;
    if (!isSubmitting) {
      socialLogin(provider);
    }
  }

  render() {
    const { email, password, message, isSubmitting, loginSuccess } = this.props;
    const { errors } = this.state;
    if (loginSuccess) {
      return this.handleNavigation("");
    }
    return (
      <div className="auth" data-test="login">
        <div className="row">
          <div className="col-md-6 hide-sm">
            <div className="auth-left" data-test="auth-left">
              <div className="logo">
                <Logo data-test="logo" />
              </div>
              <p>Join the community of like minded author, today!</p>
              <BasicButton
                className="btn btn-block"
                onClick={() => this.handleNavigation()}
                title="Sign Up"
                data-test="nav-link"
              />
            </div>
          </div>
          <div className="col-md-6 col-sm-12">
            <div className="auth-right" data-test="auth-right">
              <div className="auth-form">
                <div className="hide-md logo">
                  <Logo data-test="logo" />
                </div>
                <p className="pg-title">Sign in with</p>
                <div className="socials" data-test="socials">
                  <SocialButton
                    iconName="fb"
                    alt="fb"
                    onClick={() => this.handleSocialAuth("facebook")}
                  />
                  <SocialButton
                    iconName="twitter"
                    alt="twitter"
                    onClick={() => this.handleSocialAuth("twitter")}
                  />
                  <SocialButton
                    iconName="google-plus"
                    alt="G"
                    onClick={() => this.handleSocialAuth("google")}
                  />
                </div>
                <p className="pg-title">Or</p>
                <form data-test="login-form">
                  <TextInput
                    type="email"
                    name="email"
                    id="auth-email"
                    placeholder="Email"
                    onChange={e => this.handleOnChange(e)}
                    value={email}
                  />
                  <TextInput
                    type="password"
                    name="password"
                    id="auth-password"
                    placeholder="Password"
                    onChange={e => this.handleOnChange(e)}
                    value={password}
                  />
                  <div className="auth-errors">
                    <p className="danger">
                      {message || errors.email || errors.password}
                    </p>
                  </div>
                  <div className="">
                    <p className="password-reset" data-test="nav-link">
                      <Link to="/reset_password">Forgot password?</Link>
                    </p>
                  </div>
                  <FormButton
                    value="Sign In"
                    disabled={isSubmitting}
                    onClick={this.handleSubmit}
                  />
                </form>
                <div className="auth-link hide-md">
                  <p className="sign-up-link">
                    Don &apos; t have an account? <Link to="/">sign up</Link>.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  email: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  handleSignIn: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  handleTextInput: PropTypes.func.isRequired,
  message: PropTypes.string,
  loginSuccess: PropTypes.bool.isRequired,
  socialAuth: PropTypes.func.isRequired
};
Login.defaultProps = {
  message: ""
};
export const mapStateToProps = state => {
  const { login } = state;
  return {
    ...login,
    message: login.errors.message
  };
};

export default connect(
  mapStateToProps,
  { handleTextInput, handleSignIn, socialAuth }
)(Login);
