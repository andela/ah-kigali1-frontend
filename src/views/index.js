import React from "react";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import Login from "./Login";
import ResetPassword from "./ResetPassword";
import Home from "./Home";
import UpdatePassword from "./UpdatePassword";
import SocialAuth from "./SocialAuth";
import NewArticle from "./NewArticle";
import EditArticle from "./EditArticle";
import NotFound from "./NotFound";
import ReadArticle from "./ReadArticle";

const Routers = () => (
  <Router>
    <Switch>
      <Route exact path="/" component={Home} />
      <Route exact path="/sign_in" component={Login} />
      <Route exact path="/social_auth" component={SocialAuth} />
      <Route exact path="/reset_password" component={ResetPassword} />
      <Route exact path="/update_password" component={UpdatePassword} />
      <Route exact path="/articles/:slug/edit" component={EditArticle} />
      <Route exact path="/articles/new_article" component={NewArticle} />
      <Route exact path="/not_found" component={NotFound} />
      <Route exact path="/articles/:slug" component={ReadArticle} />
    </Switch>
  </Router>
);
export default Routers;
