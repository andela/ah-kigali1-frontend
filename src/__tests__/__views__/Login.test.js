import React from "react";
import { shallow } from "enzyme";
import sinon from "sinon";
import { Login } from "../../views/Login";
import Validator from "../../utils/validator";
import TextInput from "../../components/common/Inputs/TextInput";
import FormButton from "../../components/common/Buttons/FormButton";
import BasicButton from "../../components/common/Buttons/BasicButton";
import SocialButton from "../../components/common/Buttons/SocialButton";

const [handleTextInput, handleSignIn] = new Array(2).fill(jest.fn());
jest.mock("../../utils/validator");

const setUp = () => {
  const props = {
    isSubmitting: false,
    email: "",
    password: "",
    errors: {},
    successMessage: null,
    token: null,
    handleSignIn,
    handleTextInput
  };
  const wrapper = shallow(<Login {...props} />);
  return { props, wrapper };
};
describe("Login component", () => {
  let component;
  beforeEach(() => {
    sinon.spy(Login.prototype, "handleOnChange");
    sinon.spy(Login.prototype, "handleSubmit");
    sinon.spy(Login.prototype, "handleNavigation");
    const { wrapper } = setUp();
    component = wrapper;
  });
  afterEach(() => {
    Login.prototype.handleOnChange.restore();
    Login.prototype.handleSubmit.restore();
    Login.prototype.handleNavigation.restore();
  });
  describe("component render", () => {
    it("should render without an error", () => {
      expect(component.find(`[data-test="login"]`).length).toBe(1);
      expect(component).toMatchSnapshot();
    });
    it("should render two part of the screen", () => {
      expect(component.find(`[data-test="auth-left"]`).length).toBe(1);
      expect(component.find(`[data-test="auth-right"]`).length).toBe(1);
    });
    it("should render all basic components", () => {
      expect(component.find(`[data-test="logo"]`).length).toBe(2);
      expect(component.find(`[data-test="login-form"]`).length).toBe(1);
      expect(component.find(`[data-test="nav-link"]`).length).toBe(2);
      expect(component.find(SocialButton).length).toBe(3);
    });
    it("should render two TextInputs and button", () => {
      expect(component.find(TextInput).length).toBe(2);
      expect(component.find(FormButton).length).toBe(1);
    });
    it("should navigate to sign up page", () => {
      component
        .find(BasicButton)
        .at(0)
        .simulate("click");
      expect(Login.prototype.handleNavigation.calledOnce).toBe(true);
    });
    it("should render component with initial props", () => {
      component.find(TextInput).forEach(input => {
        expect(input.props().value).toBe("");
      });
    });
    it("should call handleOnChange when user enter input", () => {
      component
        .find(TextInput)
        .at(0)
        .simulate("change", {
          target: {
            name: "email",
            value: "luc.bayo@gmail.com"
          }
        });

      expect(Login.prototype.handleOnChange.calledOnce).toBe(true);
      expect(handleTextInput).toBeCalledWith("email", "luc.bayo@gmail.com");
    });
    it("should update password input value", () => {
      component
        .find(TextInput)
        .at(1)
        .simulate("change", {
          target: {
            name: "password",
            value: "password"
          }
        });
      expect(Login.prototype.handleOnChange.calledOnce).toBe(true);
      expect(handleTextInput).toBeCalledWith("password", "password");
    });
    it("returns validation errors", () => {
      const staticFormData = jest.fn();
      staticFormData.mockReturnValue({
        email: "Email is required"
      });
      Validator.formData = staticFormData.bind(Validator);
      component
        .find(FormButton)
        .at(0)
        .simulate("click");
      expect(Login.prototype.handleSubmit.calledOnce).toBe(true);
      expect(staticFormData).toBeCalledWith({ email: "", password: "" });
    });
    it("should call handleSubmit", () => {
      const staticFormData = jest.fn();
      staticFormData.mockReturnValue({});
      Validator.formData = staticFormData.bind(Validator);
      component
        .find(FormButton)
        .at(0)
        .simulate("click");
      expect(Login.prototype.handleSubmit.calledOnce).toBe(true);
      expect(staticFormData).toBeCalledWith({
        email: "",
        password: ""
      });
    });
  });
});
