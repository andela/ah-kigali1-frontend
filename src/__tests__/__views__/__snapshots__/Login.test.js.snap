// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component component snapshot should match the right snapshot 1`] = `
<div
  className="auth"
  data-test="login"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 hide-sm"
    >
      <div
        className="auth-left"
        data-test="auth-left"
      >
        <div
          className="logo"
        >
          <Logo
            className="logo"
            data-test="logo"
          />
        </div>
        <p>
          Join the community of like minded author, today!
        </p>
        <BasicButton
          className="btn btn-block"
          data-test="nav-link"
          onClick={[Function]}
          title="Sign Up"
        />
      </div>
    </div>
    <div
      className="col-md-6 col-sm-12"
    >
      <div
        className="auth-right"
        data-test="auth-right"
      >
        <div
          className="auth-form"
        >
          <div
            className="hide-md logo"
          >
            <Logo
              className="logo"
              data-test="logo"
            />
          </div>
          <p
            className="pg-title"
          >
            Sign in with
          </p>
          <div
            className="socials"
            data-test="socials"
          >
            <SocialButton
              alt="fb"
              className=""
              iconName="fb"
              onClick={[Function]}
            />
            <SocialButton
              alt="twitter"
              className=""
              iconName="twitter"
              onClick={[Function]}
            />
            <SocialButton
              alt="G"
              className=""
              iconName="google-plus"
              onClick={[Function]}
            />
          </div>
          <p
            className="pg-title"
          >
            Or
          </p>
          <form
            data-test="login-form"
          >
            <TextInput
              id="auth-email"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              type="email"
              value=""
            />
            <TextInput
              id="auth-password"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
            <div
              className="auth-errors"
            >
              <p
                className="danger"
              />
            </div>
            <div
              className=""
            >
              <p
                className="password-reset"
                data-test="nav-link"
              >
                <Link
                  to="/reset_password"
                >
                  Forgot password?
                </Link>
              </p>
            </div>
            <FormButton
              disabled={false}
              onClick={[Function]}
              value="Sign In"
            />
          </form>
          <div
            className="auth-link hide-md"
          >
            <p
              className="sign-up-link"
            >
              Don ' t have an account? 
              <Link
                to="/"
              >
                sign up
              </Link>
              .
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
